apm-agent-plugins/apm-apache-httpclient-plugin/src/main/java/co/elastic/apm/agent/httpclient/helper/HttpAsyncRequestProducerWrapper.java

```diff
@@ -58,26 +85,39 @@ public HttpHost getTarget() {
 
     @Override
     public HttpRequest generateRequest() throws IOException, HttpException {
-        // first read the volatile, span will become visible as a result
+        // first read the volatile, span and parent will become visible as a result
         HttpRequest request = delegate.generateRequest();
 
+        // trace context propagation
         if (request != null) {
-            RequestLine requestLine = request.getRequestLine();
-            if (requestLine != null) {
-                String method = requestLine.getMethod();
-                span.withName(method).appendToName(" ");
-                span.getContext().getHttp().withMethod(method).withUrl(requestLine.getUri());
+            if (span != null) {
+                RequestLine requestLine = request.getRequestLine();
+                if (requestLine != null) {
+                    String method = requestLine.getMethod();
+                    span.withName(method).appendToName(" ");
+                    span.getContext().getHttp().withMethod(method).withUrl(requestLine.getUri());
+                }
+                span.propagateTraceContext(request, headerSetter);
+            } else if (parent != null) {
+                parent.propagateTraceContext(request, headerSetter);
```

Comment:
[question] does the case where `span != null` is the one where an HTTP client is used directly and `parent != null` when an high level client on top of HTTP client like ES client is used ? If so, why are mutually exclusive ? 

For example, what happens if the parent span is an internal app span that does not allows to deal with any context propagation headers ? In the context of ES high-level client, both might set HTTP headers, but only the child span should propagate the context.

Reply:
Sorry, I missed that as well. 
> what happens if the parent span is an internal app span that does not allows to deal with any context propagation headers ?

I don't know what "span that does not allows to deal with any context propagation headers" means.

Let's assume the parent span set context headers based on its trace context and then this HTTP client is invoked. If it decides to create a span - it overrides the context headers with its own (as it should). If it decides not to create a span, it overrides the headers with ones with the same value, which sounds right as well.

Sorry if I didn't get what you had in mind...